Guide to get NFTs Owned by an Account (#1631)
iotaledger/iota-wiki: docs/build/isc/v1.1/docs/how-tos/core-contracts/nft/get-L2-nfts.mdx

---

> description: How to Get the L2 NFTs owned by an account.

Reviewer: ```suggestion
description: How to get the L2 NFTs owned by an account.
```

image: /img/logo/WASP_logo_dark.png
tags:
  - EVM
  - how-to
  - native tokens
  - L2 NFTs
---
import GetNftMetadata from '../../../_partials/how-tos/token/_get-nft-metadata.md';

# Get NFTs Owned by an Account

> This guide will show you how to get the L2 NFTs owned by an account using the `getL2NFTs` function from the `ISCAccounts` magic contract.

Reviewer: ```suggestion
This guide will show you how to get the L2 NFTs owned by an account using the [`ISCAccounts.getL2NFTs`](../../../reference/magic-contract/ISCAccounts.md#getl2nfts) function from the [`ISCAccounts.getL2NFTs`](../../../reference/magic-contract/ISCAccounts.md) [magic contract](../../../reference/magic-contract/introduction.md).
```

Author: Thanks for the suggestion, done.

<GetNftMetadata />

## Understanding the `getL2NFTs` Function

The `getL2NFTs` function is a view function provided by the `ISCAccounts` interface. It takes an `ISCAgentID` as an argument and returns an array of `NFTID` objects. This function allows users to query and retrieve the list of L2 NFTs owned by a specified agent. The `ISCAgentID` represents the identifier of the agent (user or contract) whose NFTs you want to retrieve. The returned array contains the identifiers of the NFTs, which can then be used to fetch more details about each NFT if needed.

> ## Why Would a User Need to retrieve owned NFTs?

Reviewer: ```suggestion
## Why Would a User Need to Retrieve Owned NFTs?```

Retrieving the NFTs owned by a user can serve several purposes:

> 1. **Portfolio Management:** Users can manage their digital assets effectively by viewing all their NFTs in one place.
> 2. **Marketplace Listing:** Users can list their NFTs on various marketplaces by verifying their ownership.
> 3. **Collection Display:** Users can showcase their NFT collections on social media, personal websites, or within specific applications.
> 4. **Trading and Transfers:** Knowing the NFTs owned helps users decide which assets to trade or transfer to other users.
> 5. **Application Integration:** Applications can provide customized experiences based on the NFTs a user owns, such as granting access to special content or features.

Reviewer: ```suggestion
### Portfolio Management

Users can manage their digital assets effectively by viewing all their NFTs in one place.

### Marketplace Listing

Users can list their NFTs on various marketplaces by verifying their ownership.

### Collection Display

Users can showcase their NFT collections on social media, personal websites, or within specific applications.

### Trading and Transfers

Knowing the NFTs owned helps users decide which assets to trade or transfer to other users.

### Application Integration

Applications can provide customized experiences based on the NFTs a user owns, such as granting access to special content or features.

```

Alternatively, make this an unordered list, as it's not instructions or a ranking.

Author: Ok, great. Please look at how I implemented this based on your alternative suggestion.

## Prerequisites

> 1. **Install the ISC Magic Library:**

Reviewer: ```suggestion

### Install the ISC Magic Library

```

>    Ensure you have the ISC Magic library installed in your project:
>    ```bash
>    npm install @iota/iscmagic
>    ```

Reviewer: ````suggestion
Ensure you have the ISC Magic library installed in your project:

```bash
npm install @iota/iscmagic
```
````

2. **Import the ISC Library:**

   In your Solidity file, import the ISC Library:
   ```solidity

>    pragma solidity >=0.8.5;

Reviewer 2: Thanks for limiting the contract version!

Author: Thanks, I'll implement this in my next commit, after receiving any suggestions from others. I appreciate your time.

> 2. **Import the ISC Library:**
> 
>    In your Solidity file, import the ISC Library:
>    ```solidity
>    pragma solidity >=0.8.5;
> 
>    import "@iota/iscmagic/ISC.sol";
>    ```

Reviewer: ````suggestion
### 2. Import the ISC Library

In your Solidity file, import the ISC Library:
```solidity
pragma solidity >=0.8.24;

import "@iota/iscmagic/ISC.sol";
```

````

## Implement the `getL2NFTs` Function

Here’s a sample implementation to retrieve L2 NFTs owned by an account:

```solidity
function getOwnedNFTs(ISCAgentID memory agentID) public view returns (NFTID[] memory) {
    // Call the getL2NFTs function from the ISCAccounts contract
    NFTID[] memory ownedNFTs = ISC.accounts.getL2NFTs(agenID);
    return ownedNFTs;
}
```

## Full Example Contract

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.5;

import "@iota/iscmagic/ISC.sol";

contract MyNFTContract {

function getOwnedNFTs(ISCAgentID memory agentID) public view returns (NFTID[] memory) {
    // Call the getL2NFTs function from the ISCAccounts contract
    NFTID[] memory ownedNFTs = ISC.accounts.getL2NFTs(agenID);
    return ownedNFTs;
    }
}
```
## Conclusion

> Using the `getL2NFTs` function from the `ISCAccounts` interface provides an efficient way to retrieve and manage the Layer 2 NFTs owned by an account. By integrating this functionality into your smart contracts, you can enhance user experience by enabling features such as portfolio management, marketplace listings, collection displays, and more. With this guide, you now have the foundational knowledge to implement and utilize the `getL2NFTs` function in your own projects, allowing for seamless interaction with `ISC Magic` and its associated NFTs.

Reviewer: ```suggestion
Using the `getL2NFTs` function from the `ISCAccounts` interface provides an efficient way to retrieve and manage the Layer 2 NFTs owned by an account. By integrating this functionality into your smart contracts, you can enhance user experience by enabling features such as [portfolio management](#portfolio-management), [marketplace listings](#marketplace-listing), [collection displays](#collection-display), and more. With this guide, you have the foundational knowledge to implement and utilize the `getL2NFTs` function in your projects, allowing for seamless interaction with `ISC Magic` and its associated NFTs.
```

---
description: How to get the L2 NFTs owned by an account.
image: /img/logo/WASP_logo_dark.png
tags:
  - EVM
  - how-to
  - native tokens
  - L2 NFTs
---
import GetNftMetadata from '../../../_partials/how-tos/token/_get-nft-metadata.md';

# Get NFTs Owned by an Account

> This guide will show you how to get the L2 NFTs owned by an account using the [`ISCAccounts.getL2NFTs`](../../../reference/magic-contract/ISCAccounts.md#getl2nfts) function from the [`ISCAccounts.getL2NFTs`](../../../reference/magic-contract/ISCAccounts.md) [magic contract](../../../reference/magic-contract/introduction.md).

Author: Thanks, updated.

<GetNftMetadata />

## Understanding the `getL2NFTs` Function

> The `getL2NFTs` function is a view function provided by the `ISCAccounts` interface. It takes an `ISCAgentID` as an argument and returns an array of `NFTID` objects. This function allows users to query and retrieve the list of L2 NFTs owned by a specified agent. The `ISCAgentID` represents the identifier of the agent (user or contract) whose NFTs you want to retrieve. The returned array contains the identifiers of the NFTs, which can then be used to fetch more details about each NFT if needed.

Reviewer 3: Maybe we should add an admonition here on what an AgentID exactly is and link to more content.

We should probably also add How-Tos on how to do the most common type conversions. Like address to agentid or NFTID to token and so on

Author: @Dr-Electron, I added an admonition, but I don't know if the info about ISCAgentID is enough. I think the URL I referenced in the admonition might not be the best.

Author: Or maybe you want me to remove this enter section and just leave the admonition?

Reviewer 3: It is good enough for now. Lets see if we want to change it later or extend the linked article in another iteration

> ## Why Would a User Need to Retrieve Owned NFTs?
> 
> Retrieving the NFTs owned by a user can serve several purposes:
> 
> - ### Portfolio Management
> 
>   - Users can manage their digital assets effectively by viewing all their NFTs in one place.
> 
> - ### Marketplace Listing
> 
>   - Users can list their NFTs on various marketplaces by verifying their ownership.
> 
> - ### Collection Display
> 
>   - Users can showcase their NFT collections on social media, personal websites, or within specific applications.
> 
> - ### Trading and Transfers
> 
>   - Knowing the NFTs owned helps users decide which assets to trade or transfer to other users.
> 
> - ### Application Integration
> 
>   - Applications can provide customized experiences based on the NFTs a user owns, such as granting access to special content or features.

Reviewer 3: ```suggestion
```

I would keep the how-tos on point. Not too much stuff around it. We can create user stories in tutorials imo

Author: Ok, removed.

> ## Prerequisites
> 
> ### 1. Install the ISC Magic Library
> 
>    Ensure you have the ISC Magic library installed in your project:
> 
>    ```bash
>    npm install @iota/iscmagic
>    ```
> 
> ### 2. Import the ISC Library
> 
>    In your Solidity file, import the ISC Library:
> 
>    ```solidity
>    pragma solidity >=0.8.24;
> 
>    import "@iota/iscmagic/ISC.sol";
>    ```

Reviewer 3: Let's not duplicate content. We explain this I the magic contract introduction
```suggestion
```

## Implement the `getL2NFTs` Function

Here’s a sample implementation to retrieve L2 NFTs owned by an account:

```solidity
function getOwnedNFTs(ISCAgentID memory agentID) public view returns (NFTID[] memory) {
    // Call the getL2NFTs function from the ISCAccounts contract
    NFTID[] memory ownedNFTs = ISC.accounts.getL2NFTs(agenID);
    return ownedNFTs;
}
```

## Full Example Contract

```solidity
// SPDX-License-Identifier: MIT
pragma solidity >=0.8.24;

import "@iota/iscmagic/ISC.sol";

contract MyNFTContract {

function getOwnedNFTs(ISCAgentID memory agentID) public view returns (NFTID[] memory) {
    // Call the getL2NFTs function from the ISCAccounts contract
    NFTID[] memory ownedNFTs = ISC.accounts.getL2NFTs(agenID);
    return ownedNFTs;
    }
}
```

> ## Conclusion

Reviewer 3: I like consistency, so I would remove it as we don't have it in other how-tos either. But no strong opinion from my side. Question is if someone will read it after the full how-to
